test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go

```diff
@@ -0,0 +1,508 @@
+			&handler.EnqueueRequestsFromMapFunc{
+				ToRequests: utilexp.MachinePoolToInfrastructureMapFunc(infrav1exp.GroupVersion.WithKind("DockerMachinePool"), r.Log),
+			},
+		).
+		Build(r)
+	if err != nil {
+		return err
+	}
+	return c.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: clusterToDockerMachinePools,
+		},
+		predicates.ClusterUnpausedAndInfrastructureReady(r.Log),
+	)
+}
+
+func (r *DockerMachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinePool *clusterv1exp.MachinePool, dockerMachinePool *infrav1exp.DockerMachinePool, log logr.Logger) (ctrl.Result, error) {
+	pool, err := newNodePool(r, cluster, machinePool, dockerMachinePool, log)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to build new node pool")
+	}
+
+	if err := pool.Delete(ctx); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to delete all machines in the node pool")
+	}
+
+	controllerutil.RemoveFinalizer(dockerMachinePool, infrav1exp.MachinePoolFinalizer)
+	return ctrl.Result{}, nil
+}
+
+func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, cluster *clusterv1.Cluster, machinePool *clusterv1exp.MachinePool, dockerMachinePool *infrav1exp.DockerMachinePool, log logr.Logger) (ctrl.Result, error) {
+	// Make sure bootstrap data is available and populated.
+	if machinePool.Spec.Template.Spec.Bootstrap.DataSecretName == nil {
+		log.Info("Waiting for the Bootstrap provider controller to set bootstrap data")
+		return ctrl.Result{}, nil
+	}
+
+	if machinePool.Spec.Replicas == nil {
+		machinePool.Spec.Replicas = pointer.Int32Ptr(1)
+	}
+
+	pool, err := newNodePool(r, cluster, machinePool, dockerMachinePool, log)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to build new node pool")
+	}
+
+	// if we don't have enough nodes matching spec, build them
+	if err := pool.ReconcileMachines(ctx); err != nil {
+		if errors.Is(err, &TransientError{}) {
+			log.V(4).Info("requeue in 5 seconds due docker machine reconcile transient error")
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+
+		return ctrl.Result{}, errors.Wrap(err, "failed to reconcile machines")
+	}
+
+	if err := pool.DeleteExtraMachines(ctx); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to delete overprovisioned or out of spec machines")
+	}
+
+	dockerMachinePool.Spec.ProviderIDList = []string{}
+	for _, instance := range dockerMachinePool.Status.Instances {
+		if instance.ProviderID != nil && instance.Ready {
+			dockerMachinePool.Spec.ProviderIDList = append(dockerMachinePool.Spec.ProviderIDList, *instance.ProviderID)
+		}
+	}
+
+	dockerMachinePool.Status.Replicas = int32(len(dockerMachinePool.Spec.ProviderIDList))
+
+	if dockerMachinePool.Spec.ProviderID == "" {
+		dockerMachinePool.Spec.ProviderID = getDockerMachinePoolProviderID(cluster.Name, dockerMachinePool.Name)
```

Comment:
might be worth addding a comment that this is a fake id given that in docker the is no a native abstraction grouping machiens like in Azure or other cloud provider